ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -0,0 +1,489 @@
+    sample *queued_sample = &heap_recorder->queued_samples[i];
+    if (!queued_sample->skip) {
+      fprintf(stderr, "Flushing sample %p\n", queued_sample);
+
+      if (queued_sample->free) {
+        commit_free(heap_recorder, queued_sample->obj);
+      } else {
+        commit_allocation(heap_recorder, queued_sample->stack, queued_sample->obj, queued_sample->weight);
+      }
+    }
+
+    *queued_sample = EmptySample;
+  }
+  heap_recorder->queued_samples_len = 0;
+}
+
+static void enqueue_sample(heap_recorder *heap_recorder, sample new_sample) {
+  fprintf(stderr, "Enqueuing sample for %lu (weight=%u free=%i)\n", new_sample.obj, new_sample.weight, new_sample.free);
+  if (heap_recorder->queued_samples_len >= MAX_QUEUE_LIMIT) {
+    fprintf(stderr, "Dropping sample on the floor.\n");
+    return;
+  }
+
+  heap_recorder->queued_samples[heap_recorder->queued_samples_len] = new_sample;
+  heap_recorder->queued_samples_len++;
+}
+
+static void enqueue_allocation(heap_recorder *heap_recorder, heap_stack *heap_stack, VALUE obj, unsigned int weight) {
+  enqueue_sample(heap_recorder, (sample) {
+      .stack = heap_stack,
+      .obj = obj,
+      .weight = weight,
+      .free = false,
+      .skip = false,
+  });
+}
+
+static void enqueue_free(heap_recorder *heap_recorder, VALUE obj) {
+  enqueue_sample(heap_recorder, (sample) {
+      .stack = NULL,
+      .obj = obj,
+      .weight = 0,
+      .free = true,
+      .skip = false,
+  });
+}
+
+void start_heap_allocation_recording(heap_recorder* heap_recorder, VALUE new_obj, unsigned int weight, ddog_CharSlice *class_name) {
+  fprintf(stderr, "Started recording allocation of %lu with weight %u\n", new_obj, weight);
+  partial_heap_recording *active_recording = &heap_recorder->active_recording;
+  active_recording->obj = new_obj;
+  active_recording->weight = weight;
+  active_recording->class_name = class_name;
+}
+
+void end_heap_allocation_recording(struct heap_recorder *heap_recorder, ddog_prof_Slice_Location locations) {
+  // TODO: Make use of active_recording->class_name
+  partial_heap_recording *active_recording = &heap_recorder->active_recording;
+
+  VALUE new_obj = active_recording->obj;
+  if (!new_obj) {
+    // Recording ended without having been started?
+    rb_raise(rb_eRuntimeError, "Ended a heap recording that was not started");
+    return;
+  }
+
+  int weight = active_recording->weight;
+
+  // From now on, mark active recording as invalid so we can short-circuit at any point and
+  // not end up with a still active recording. new_obj still holds the object for this recording
+  active_recording->obj = 0;
+
+  heap_stack *heap_stack = heap_stack_init(locations);
+  int error = pthread_mutex_trylock(&heap_recorder->records_mutex);
+  if (!error) {
+    // We were able to get a lock to heap_records so lets flush any pending samples
+    // that might have been queued previously before adding this new one.
+    flush_queue(heap_recorder);
```

Comment:
Minor: More of a style thing, but the current structure reads a bit weird to me

```
error = ...
if (!error) {
  happy path steps
} else {
  error fallback
  return
}

more happy path steps
```

I'd suggest doing `if (error) { .... return; }` and keeping the happy path all together.